tls/s2n_cipher_suites.c

```diff
@@ -1331,6 +1332,17 @@ static int s2n_set_cipher_as_server(struct s2n_connection *conn, uint8_t *wire,
                 continue;
             }
 
+            /* The s2n server should not select ecdhe ciphersuite if supported_groups is not sent by client.
+             * The supported_groups extension indicates the elliptic curves supported by client.The negotiated_curve
+             * being NULL indicates that no curve was agreed upon during key exchange,and the communication is not secured
+             * using elliptic curve cryptography. ecdhe ciphersuite requires an elliptic curve to generate keys and hence cannot
+             * be selected when supported_groups extension is not sent. */
+            if (conn->kex_params.server_ecc_evp_params.negotiated_curve == NULL) {
```

Comment:
The test you wrote succeeds without this change. I looked at the code and it seems like we're already checking for a negotiated curve before allowing an ECDHE cipher: https://github.com/aws/s2n-tls/blob/8db1211d790033a088854e5d8686eb1a9074915c/tls/s2n_cipher_suites.c#L1290
https://github.com/aws/s2n-tls/blob/8db1211d790033a088854e5d8686eb1a9074915c/tls/s2n_kex.c#L57

I tested the scenario presented in https://github.com/aws/s2n-tls/issues/2070 by building a s2nc that doesn't send a supported groups extension, and was able to replicate it. It looks like an ECDHE cipher is able to be negotiated because we set a default curve when receiving a client hello, so the negotiated curve will never be null: https://github.com/aws/s2n-tls/blob/8db1211d790033a088854e5d8686eb1a9074915c/tls/s2n_client_hello.c#L455-L462

Since the client is advertising ECDHE ciphers, the current behavior of picking a default curve doesn't seem too unreasonable to me. But, I checked OpenSSL and it looks like they *don't* negotiate ECDHE without the supported_groups extension.

So, I'm not sure if we want to keep the existing behavior or remove the default curve for < TLS 1.3.